server/backends/disk_cache/disk_cache.go

```diff
@@ -597,49 +588,158 @@ func (fk *fileKey) FromPartitionAndPath(part *partition, fullPath string) error
 
 	// pull group off the front
 	if len(parts) > 0 {
-		fk.userPrefix = fk.part.internString(parts[0])
+		userPrefix = parts[0]
 		parts = parts[1:]
 	} else {
-		return parseError()
+		err = parseError()
+		return
 	}
 
 	// pull digest off the end
 	if len(parts) > 0 {
-		digestBytes, err := hex.DecodeString(parts[len(parts)-1])
-		if err != nil {
-			return parseError()
+		db, decodeErr := hex.DecodeString(parts[len(parts)-1])
+		if decodeErr != nil {
+			err = parseError()
+			return
 		}
-		fk.digestBytes = digestBytes
+		digestBytes = db
 		parts = parts[:len(parts)-1]
 	} else {
-		return parseError()
+		err = parseError()
+		return
 	}
 
 	// If this is a v2-formatted path, remove the hash prefix dir
-	if fk.part.useV2Layout {
+	if useV2Layout {
 		if len(parts) > 0 {
 			prefixPart := parts[len(parts)-1]
 			if len(prefixPart) != HashPrefixDirPrefixLen {
-				return parseError()
+				err = parseError()
+				return
 			}
 			parts = parts[:len(parts)-1]
 		} else {
-			return parseError()
+			err = parseError()
+			return
 		}
 	}
 
 	// If this is an /ac/ directory, this was an actionCache item
 	// otherwise it was a CAS item.
 	if len(parts) > 0 && parts[len(parts)-1] == "ac" {
-		fk.cacheType = interfaces.ActionCacheType
+		cacheType = interfaces.ActionCacheType
 		parts = parts[:len(parts)-1]
 	} else {
-		fk.cacheType = interfaces.CASCacheType
+		cacheType = interfaces.CASCacheType
 	}
 
 	if len(parts) > 0 {
-		fk.remoteInstanceName = fk.part.internString(strings.Join(parts, "/"))
+		remoteInstanceName = strings.Join(parts, "/")
 	}
+	return
+}
+
+func ScanDiskDirectory(scanDir string) <-chan *rfpb.FileMetadata {
+	scanned := make(chan *rfpb.FileMetadata)
```

Comment:
I think you want this to be a bufferred channel for better throughput.

Reply:
Done, thanks!